tag.go

```diff
@@ -1,136 +1,174 @@
+			out.TagType = _String{string(line[6:])}
+		case bytes.HasPrefix(line, []byte(" tag ")):
+			out.Tag = _String{string(line[5:])}
+		case bytes.HasPrefix(line, []byte(" tagger ")):
+			tagger, err := parsePersonInfo(line[1:])
+			if err != nil {
+				return nil, nil, err
+			}
+			out.Tagger = *tagger
+		case string(line) == " ":
+			for {
+				line, _, err := rd.ReadLine()
+				if err != nil {
+					return nil, nil, err
+				}
+
+				if !bytes.HasPrefix(line, []byte(" ")) {
+					return &out, line, nil
+				}
+
+				out.Text.x += string(line) + "\n"
+			}
 		}
-		return t.Tagger.resolve(path[1:])
-	case "message":
-		return t.Message, path[1:], nil
-	case "tag":
-		return t.Tag, path[1:], nil
-	default:
-		return nil, nil, errors.New("no such link")
 	}
+	return &out, nil, nil
 }
 
-func (t *Tag) ResolveLink(path []string) (*node.Link, []string, error) {
-	out, rest, err := t.Resolve(path)
+func encodeTag(n ipld.Node, w io.Writer) error {
+	obj, err := n.LookupByString("Object")
 	if err != nil {
-		return nil, nil, err
+		return err
 	}
-
-	lnk, ok := out.(*node.Link)
-	if !ok {
-		return nil, nil, errors.New("not a link")
+	objLnk, err := obj.AsLink()
+	if err != nil {
+		return err
 	}
 
-	return lnk, rest, nil
-}
+	tt, err := n.LookupByString("TagType")
+	if err != nil {
+		return err
+	}
+	ttStr, err := tt.AsString()
+	if err != nil {
+		return err
+	}
 
-func (t *Tag) Size() (uint64, error) {
-	return uint64(len(t.RawData())), nil
-}
+	tag, err := n.LookupByString("Tag")
+	if err != nil {
+		return err
+	}
+	tagStr, err := tag.AsString()
+	if err != nil {
+		return err
+	}
 
-func (t *Tag) Stat() (*node.NodeStat, error) {
-	return &node.NodeStat{}, nil
-}
+	text, err := n.LookupByString("Text")
+	if err != nil {
+		return err
+	}
+	textStr, err := text.AsString()
+	if err != nil {
+		return err
+	}
 
-func (t *Tag) String() string {
-	return "[git tag object]"
-}
+	tagger, taggerErr := n.LookupByString("Tagger")
 
-func (t *Tag) Tree(p string, depth int) []string {
-	if p != "" {
-		if p == "tagger" {
-			return []string{"name", "email", "date"}
+	buf := new(bytes.Buffer)
+	fmt.Fprintf(buf, "object %s\n", hex.EncodeToString(sha(objLnk)))
+	fmt.Fprintf(buf, "type %s\n", ttStr)
+	fmt.Fprintf(buf, "tag %s\n", tagStr)
+	if taggerErr == nil && !tagger.IsNull() {
+		pi := Type.PersonInfo__Repr.NewBuilder()
+		if err := pi.AssignNode(tagger); err != nil {
+			return err
 		}
-		return nil
+		piN := pi.Build()
+		parsed, ok := piN.(*_PersonInfo)
+		if !ok {
+			return fmt.Errorf("Could not parse tagger person info %v", tagger)
+		}
+		fmt.Fprintf(buf, "tagger %s\n", parsed.GitString())
 	}
-	if depth == 0 {
-		return nil
+	if textStr != "" {
+		fmt.Fprintf(buf, "\n%s", textStr)
 	}
 
-	tree := []string{"object", "type", "tag", "message"}
-	tree = append(tree, t.Tagger.tree("tagger", depth)...)
-	return tree
-}
-
-func (t *Tag) GitSha() []byte {
-	return cidToSha(t.Cid())
+	if _, err := fmt.Fprintf(w, "tag %d\x00", buf.Len()); err != nil {
```

Comment:
any particular reason you write the word-length-null directly to w, but the rest via buffer? I think you should do everything the same way, be it direct writes, a bytes.Buffer, or a bufio.Writer. It's also a bit confusing that the writes seem to happen out of order, because you fill the buffer first but flush it later.

(this applies to all other uses of buf.WriteTo too)

Reply:
i guess we first build the object, and only then can generate the prefix with the length of it, so we don't how long  of a prefix to allocate at the beginning of the buffer until the end.

Reply:
Oh, right, of course :) That also kinda disproves the point about bufio.Writer, as it doesn't keep track of written bytes.